### PR TITLE
vLLM V1 migration

### PR DESCRIPTION
**Pull request by rafapi**
Upgrade to vLLMv1 and restore vLLM V0 behaviour on the vLLM V1 rollout path
| reward | logprobs |
|:--------:|:------:|
|image|image
> The logprobs shows the discrepancy with the initial vanilla vllm upgrade (green line)

### CODE REVIEW

**Comment by rafapi** [comment]
@bigximik FYI

File: pipelinerl/finetune/rl/__init__.py [hunk 1]
```diff
@@ -199,18 +199,36 @@ def rl_step(
...
+        log_norm_unsqueezed = detached_log_norm.unsqueeze(-1)
+        vocab_chunk_size = 4096
+        with torch.no_grad():
+            for start in range(0, detached_logits.shape[-1], vocab_chunk_size):
+                chunk = detached_logits[..., start:start + vocab_chunk_size]
+                chunk_logprobs = chunk - log_norm_unsqueezed
+                chunk_probs = torch.exp(chunk_logprobs)
+                entropy -= (chunk_probs * chunk_logprobs).sum(dim=-1)
+
+    del logits, selected_logits, log_norm
```
Comment (rafapi) [inline pipelinerl/finetune/rl/__init__.py:231]: all the chamges abpve are an old bugfix to ensure we don't materialise the vocab tensors by default and taking a lot of VRAM for no reason

File: pipelinerl/vllm1.py [hunk 2]
```diff
@@ -73,6 +79,24 @@ def init_actor_update_group(
...
+            # tree algo, simple proto, P2P disabled) that the trainer does not share.
+            # Clear them so the weight-update NCCL comm matches trainer defaults.
+            # Safe at tp=1 because no intra-engine NCCL comm has been created yet.
+            for _k in (
+                "NCCL_LAUNCH_MODE", "NCCL_COLLNET_ENABLE", "NCCL_NVLS_ENABLE",
+                "NCCL_P2P_NET_DISABLE", "NCCL_MIN_NCHANNELS", "NCCL_MAX_NCHANNELS",
+                "NCCL_PROTO", "NCCL_ALGO", "NCCL_NTHREADS", "NCCL_SOCKET_NTHREADS",
+            ):
+                os.environ.pop(_k, None)
+
```
Comment (rafapi) [inline pipelinerl/vllm1.py:100]: This stuff above is not required but since i tested batch invariance i decided to leave it for completion

File: pipelinerl/async_llm.py [hunk 3]
```diff
@@ -90,27 +119,56 @@ async def llm_async_generate(
...
+            abort_reason,
+            choice.get("finish_reason"),
+            usage.get("prompt_tokens"),
+            usage.get("completion_tokens"),
+        )
+        if attempt == 0:
+            continue
+        raise RetryableAbortedCompletionError(
+            f"Repeated aborted completion for prompt {prompt.id}: {abort_reason}"
+        )
```
Comment (ehsk) [inline pipelinerl/async_llm.py]: nitpick, simplify this if:

```python
if attempt > 0:
  raise RetryableAbortedCompletionError(...)
```

File: pipelinerl/async_llm.py [hunk 4]
```diff
@@ -90,27 +119,56 @@ async def llm_async_generate(
...
-        headers=headers,
-        ssl=False,
-    ) as response:
-        if not response.ok:
-            error_text = await response.text()
-            logger.error(f"Failed to get completion: {error_text}")
-            response.raise_for_status()
-        data = await response.json()
+    response_data = None
+    for attempt in range(2):
```
Comment (ehsk) [inline pipelinerl/async_llm.py:123]: why 3 times?

File: pipelinerl/vllm1.py [hunk 5]
```diff
@@ -73,6 +80,24 @@ def init_actor_update_group(
...
+        try:
+            batch_invariant_enabled = int(batch_invariant_env) != 0
+        except ValueError:
+            batch_invariant_enabled = False
+
+        if batch_invariant_enabled:
+            # vLLM batch_invariant mode sets restrictive NCCL env vars (single channel,
+            # tree algo, simple proto, P2P disabled) that the trainer does not share.
+            # Clear them so the weight-update NCCL comm matches trainer defaults.
+            # Safe at tp=1 because no intra-engine NCCL comm has been created yet.
```
Comment (ehsk) [inline pipelinerl/vllm1.py:93]: Shouldn't we add an assert to ensure this should occur only when tensor-parallel-size is 1?

File: pipelinerl/async_llm.py (hunk 4 above)
Comment (rafapi) [reply pipelinerl/async_llm.py:123]: twice :), just a basic retry

File: pipelinerl/vllm1.py [hunk 6]
```diff
@@ -127,11 +153,36 @@ async def input_process_groups(self):
...
 
     async def receive_weight_update(self, request: WeightUpdateRequest):
-        logger.info("Starting weight update...")
-        await self.engine_client.collective_rpc_async(
-            "receive_weight_update", args=(request.model_dump_json(),)
-        )
-        logger.info("Weight update processed")
+        async with self.update_lock:
+            version = getattr(request, "version", "unknown")
+            paused = False
```
Comment (ehsk) [inline pipelinerl/vllm1.py]: I don't get the purpose of `paused` as it's a local variable and will always be `true` in the try/finally clause.

File: pipelinerl/vllm1.py (hunk 5 above)
Comment (rafapi) [reply pipelinerl/vllm1.py:93]: see my comment in the files, this mode should never be used, i left it there because it's a useful default for NCCL compatibility with the trainer

File: pipelinerl/vllm1.py (hunk 6 above)
Comment (rafapi) [reply pipelinerl/vllm1.py]: yep, old code logic

File: pipelinerl/vllm1.py (hunk 5 above)
Comment (ehsk) [reply pipelinerl/vllm1.py:93]: Do we know if this works when batch invariant is enabled and tp > 1?

File: pipelinerl/async_llm.py (hunk 4 above)
Comment (ehsk) [reply pipelinerl/async_llm.py:123]: my bad ;) can we make it configurable? and add a small delay between the retries?

Comment (rafapi) [reply pipelinerl/async_llm.py:123]: No, I’d keep this as a single immediate retry. This path should stay as fast as possible, and if these aborts happen at all, we don’t want to add extra delay here.

File: pipelinerl/vllm1.py (hunk 5 above)
Comment (rafapi) [reply pipelinerl/vllm1.py:93]: it makes things slightly worse, it adds about 2x delay that compounds through the run

File: pipelinerl/async_llm.py (hunk 4 above)
Comment (ehsk) [reply pipelinerl/async_llm.py:123]: This makes me wonder what specifically causes these aborts in this system. If it's weight-update pauses, an immediate retry might just hit another abort if the update hasn't finished yet. In that case, the retry only helps if the pause is shorter than the round-trip time of the HTTP request, which seems fragile to assume.

File: pipelinerl/vllm1.py (hunk 5 above)
Comment (ehsk) [reply pipelinerl/vllm1.py:93]: Great, can we also clarify it in the comments for tp > 1?

File: pipelinerl/async_llm.py (hunk 4 above)
Comment (rafapi) [reply pipelinerl/async_llm.py:123]: To be fair, this is a very narrow case surfaced by Fast-LLM, i haven't ever hit it on normal runs. let's keep it as a defensive mechanism for now, it won't hurt

**Review by ehsk — APPROVED** [review]
LGTM!
